plat/kvm/x86/lcpu_start.S

```diff
@@ -0,0 +1,305 @@
+
+#define CR4_BOOT32_SETTINGS						\
+	  X86_CR4_PAE	/* Physical Address Extension */
+
+#define EFER_BOOT32_SETTINGS						\
+	  X86_EFER_LME	/* IA-32e Mode */				\
+	| X86_EFER_NXE	/* Execute Disable Bit */
+
+#define CR0_BOOT32_SETTINGS						\
+	  X86_CR0_PE	/* Protected mode */				\
+	| X86_CR0_WP	/* Write Protect */				\
+	| X86_CR0_PG	/* Paging */
+
+.code32
+.section .text.boot.32
+ENTRY(lcpu_start32)
+	/* Enable physical address extension (PAE) */
+	movl	$CR4_BOOT32_SETTINGS, %eax
+	movl	%eax, %cr4
+
+	/* Switch to IA-32e mode (long mode) */
+	xorl	%edx, %edx
+	movl	$EFER_BOOT32_SETTINGS, %eax
+	movl	$X86_MSR_EFER, %ecx
+	wrmsr
+
+	/* Set boot page table and enable paging */
+	movl	$x86_bpt_pml4, %eax
+	movl	%eax, %cr3
+
+	movl	$CR0_BOOT32_SETTINGS, %eax
+	movl	%eax, %cr0
+
+	/* Load 64-bit GDT and jump to 64-bit code segment */
+	lgdt	gdt64_ptr
+	ljmp	$(gdt64_cs - gdt64), $jump_to64
+.code64
+jump_to64:
+	/* Set up remaining segment registers */
+	movl	$(gdt64_ds - gdt64), %eax
+	movl	%eax, %es
+	movl	%eax, %ss
+	movl	%eax, %ds
+
+	xorl	%eax, %eax
+	movl	%eax, %fs
+	movl	%eax, %gs
+
+	leaq	lcpu_start64, %rcx
+	jmp	*%rcx
+END(lcpu_start32)
+
+/*
+ * 64-bit boot entry function
+ * System must be in 64-bit mode with paging enabled. The first 4GiB have to be
+ * identity mapped. A GDT must be loaded with descriptors for 4GiB flat CS and
+ * DS segments. CS must have execute/read permission and DS must have
+ * read/write permission. ES and ES must be the same as DS. Interrupts must be
+ * disabled.
+ *
+ * NOTE: We assume a pointer to a struct lcpu_sargs in RDI and a pointer to
+ * potential platform boot arguments in RSI (e.g., multiboot) if this is the
+ * first boot. In that case, we setup the CPU with the provided startup
+ * arguments.
+ *
+ * NOTE: Code should be position-independent
+ */
+#include "lcpu_helpers.S"
+
+.code64
+.section .text.boot.64
+ENTRY(lcpu_start64)
+	/* Save the startup args pointer */
+	movq	%rdi, %r8
+
+	/* Request basic CPU features and APIC ID
+	 * TODO: This APIC ID is limited to 256. Better get from leaf 0x1f
+	 */
+	movl	$1, %eax
+	cpuid
+	shrl	$24, %ebx
+
+	/* Use APIC_ID * LCPU_SIZE for indexing the cpu structure */
+	movl	$LCPU_SIZE, %eax
+	imul	%ebx, %eax
+
+	/* Compute pointer into CPU struct array and store it in RBP
+	 * We do not use the frame pointer, yet
+	 */
+	leaq	lcpus(%rip), %rbp
+	addq	%rax, %rbp
+
+	/* Put CPU into init state */
+	movl	$LCPU_STATE_INIT, LCPU_STATE_OFFSET(%rbp)
+
+	/* Enable FPU and SSE */
+	LCPU_ENABLE_FPU_SSE
+
+#if (__AVX__ || CONFIG_HAVE_X86PKU)
+	/* Enable XSAVE feature */
+	LCPU_ENABLE_XSAVE fail
+#if __AVX__
+	/* Enable AVX */
+	LCPU_ENABLE_AVX fail
+#endif /* __AVX__ */
+#endif /* __AVX__ || CONFIG_HAVE_X86PKU */
+
+	/* Request extended CPU features */
+	movl	$7, %eax
+	xorl	%ecx, %ecx
+	cpuid
+
+	/* Enable FS and GS base */
+	LCPU_ENABLE_FSGSBASE no_fsgsbase
+no_fsgsbase:
+#if CONFIG_HAVE_X86PKU
+	/* Enable memory protection keys (PKU) */
+	LCPU_ENABLE_PKU fail
```

Comment:
Instead of halting the CPU you can simply proceed without memory protection keys by defining new label `no_pku`

Reply:
Actually, this is also the behavior in the current code and should be kept. Thanks for pointing out! I fixed this in PR #501. Thanks 👍🏻